### DISCUSSION TITLE
Protocol for other languages

### DISCUSSION THREAD

**Discussion by akvadrako** [appy-one/acebase#79] [General]
Is the acebase protocol documented?

I would be interested in using it from python.

**Comment by appy-one** [comment]
Not yet, but I will be working on that pretty soon!

**Comment by appy-one** [comment]
I've started documenting the server API methods using OpenAPI (Swagger) specs. This is still work in progress, but many endpoints are already documented and more will follow soon.

See the yaml files in [acebase-server/tree/master/src/routes](https://github.com/appy-one/acebase-server/tree/master/src/routes) for info about the endpoints, they will also be testable from the server's _/docs_ endpoint if the set `NODE_ENV` is `development`.

Note that most of this is currently only available on Github, I'm busy porting all server js code to typescript and not ready to publish to npm yet. Let me know if you have any questions

**Reply by akvadrako** [reply]
I looked at one of the ones I'm interested in (data-update.yaml). It seems reasonable but how to you deal with framing the requests over the websocket connection?

I guess I'm looking for more of a high-level overview of the protocol before considering individual requests. The most interesting part is how the real-time updates are handled and how partial updates are represented.

**Reply by appy-one** [reply]
The websocket is currently only used for realtime notifications and managing subscriptions thereof. The transaction endpoint being the only exception. I'll document the websocket messages too shortly, take a look at [src/websocket/manager.ts](https://github.com/appy-one/acebase-server/blob/master/src/websocket/manager.ts) for more info about the messages and the expected data

**Comment by akvadrako** [comment]
I've got some sample python code working, (https://gist.github.com/akvadrako/8ac132bfeeb5971d43772cc8b210404b). Mostly pretty straightforward; I will try to get websockets working next.

What I realize I'm missing is some kind of long-term sync / streaming solution. When I have the websocket working I'll be able to get real-time updates, but if my client disconnects there is no way to stay in sync without fetching all the state again.

**Comment by appy-one** [comment]
If you enable transaction logging in the server config, an `acebase_cursor` property is added to snapshots' `context()` data, and event data sent to your client over the websocket connection. You can use that cursor to fetch effective changes for specific path/event combinations since the last cursor you received with the [/sync/{dbname}/changes endpoint](https://github.com/appy-one/acebase-server/blob/master/src/routes/data-sync-changes.yaml)

**Reply by akvadrako** [reply]
Thanks, that works well. I've got my sample code working with the changes endpoint and the websocket. I submitted some small PRs to fix the stuff that was really in my way but mostly it was pretty smooth.
